lib/MusicBrainz/Script/RemoveUnreferencedRows.pm

```diff
@@ -0,0 +1,136 @@
+            SQL
+    };
+
+    for my $table_group (@unreferenced_rows) {
+        my ($table_name, $row_ids) = @$table_group;
+
+        log_info { sprintf "Checking rows for $table_name" };
+
+        my $quoted_table_name =
+            $self->c->sql->dbh->quote_identifier($table_name);
+
+        # We want to ensure the rows are still unreferenced
+        my $query = <<~"SQL";
+            SELECT id
+              FROM $quoted_table_name
+             WHERE ref_count = 0
+               AND id = any(?)
+            SQL
+
+        my $unused_ids = $self->c->sql->select_single_column_array(
+            $query,
+            $row_ids,
+        );
+
+        $count += scalar @$unused_ids;
+
+        for my $id (@$row_ids) {
+            if (any {/^$id$/} @$unused_ids) {
+                if ($self->dry_run) {
+                    log_info {
+                        sprintf "Need to remove id=%s from $table_name",
+                        $id,
+                    } if $self->verbose;
+                }
+                else {
+                    capture_exceptions(sub {
+                        Sql::run_in_transaction(sub {
+                            if ($table_name eq 'artist_credit') {
+                                $self->sql->do(<<~'SQL', $id);
+                                    DELETE FROM artist_credit_gid_redirect
+                                          WHERE new_id = ?
+                                    SQL
+                            }
+                            $self->c->sql->do(<<~"SQL", $id);
+                                DELETE FROM $quoted_table_name
+                                      WHERE id = ?
+                                SQL
+                            $self->c->sql->do(<<~'SQL', $table_name, $id);
+                                DELETE FROM unreferenced_row_log
+                                      WHERE table_name = ?
+                                        AND row_id = ?
+                                SQL
+                            ++$removed
+                        }, $self->c->sql);
+                    }, sub {
+                        my $err = shift;
+                        log_info { "Error while processing id $id from $table_name: $err\n" };
+                        return;
+                    });
+                }
+            } else {
+                if ($self->dry_run) {
+                    log_info {
+                        sprintf "id=%s from $table_name is no longer unreferenced",
+                        $id,
+                    } if $self->verbose;
+                } else {
+                    # If no longer unreferenced, drop from the unreferenced table
+                    Sql::run_in_transaction(sub {
+                        $self->c->sql->do(<<~'SQL', $table_name, $id);
+                            DELETE FROM unreferenced_row_log
+                                  WHERE table_name = ?
+                                    AND row_id = ?
```

Comment:
This has a different race condition in that the row can become de-referenced again in between when the `$unused_ids` query runs and now.  (Then it will never be cleaned up.)